File: csharp/EPAM.Deltix.HdTime.Tests/HdTimeSpanTest.cs
/*
  Copyright 2021 EPAM Systems, Inc

  See the NOTICE file distributed with this work for additional information
  regarding copyright ownership. Licensed under the Apache License,
  Version 2.0 (the "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
  License for the specific language governing permissions and limitations under
  the License.
 */

using System;
using System.Text;
using NUnit.Framework;


namespace EPAM.Deltix.HdTime.Tests
{
	[TestFixture]
	public class HdTimeSpanTest
	{
		private static Object[] TestDates = {new Object[] {DateTime.Now, DateTime.Now.AddDays(-1)}};
		private static Object[] TestIntervals = {new Object[] {TimeSpan.FromDays(123), TimeSpan.FromDays(12345)}};

		// These intervals were broken and don't fit into HdTimeSpan (>1900 years each)
		//private static Object[] TestIntervals = { new Object[] { TimeSpan.FromDays(DateTime.Now.Ticks), TimeSpan.FromTicks(DateTime.Now.AddDays(-1).Ticks) } };

		[Test]
		public void TestConstants()
		{
			Assert.AreEqual(HdTimeSpan.NanosInDay, 100 * TimeSpan.TicksPerDay);
			Assert.AreEqual(HdTimeSpan.NanosInHour, 100 * TimeSpan.TicksPerHour);
			Assert.AreEqual(HdTimeSpan.NanosInMillisecond, 100 * TimeSpan.TicksPerMillisecond);
			Assert.AreEqual(HdTimeSpan.NanosInMinute, 100 * TimeSpan.TicksPerMinute);
			Assert.AreEqual(HdTimeSpan.NanosInSecond, 100 * TimeSpan.TicksPerSecond);
			Assert.AreEqual(0, HdTimeSpan.Zero.TotalNanoseconds);
			Assert.AreEqual(-Int64.MaxValue, HdTimeSpan.MinValue.TotalNanoseconds);
			Assert.AreEqual(Int64.MaxValue, HdTimeSpan.MaxValue.TotalNanoseconds);
		}

		[Test]
		[TestCaseSource("TestIntervals")]
		public void TestAdd(TimeSpan a, TimeSpan b)
		{
			TimeSpan sum = a + b;

			HdTimeSpan hdA = a;
			HdTimeSpan hdB = b;
			HdTimeSpan hdSum = hdA + hdB;
			Assert.IsTrue(hdA == a);
			Assert.IsTrue(hdB == b);
			Assert.IsTrue(hdSum == sum);
		}


		[Test]
		[TestCaseSource("TestIntervals")]
		public void TestSubtract(TimeSpan a, TimeSpan b)
		{
			TimeSpan timeSpan = a - b;

			HdTimeSpan hdA = a;
			HdTimeSpan hdB = b;
			HdTimeSpan hdTimeSpan = hdA - hdB;
			Assert.IsTrue(hdTimeSpan == timeSpan);
		}


		[Test]
		[TestCaseSource("TestDates")]
		public void TestSubtractDates(DateTime a, DateTime b)
		{
			TimeSpan timeSpan = a - b;

			HdDateTime hdA = a;
			HdDateTime hdB = b;
			HdTimeSpan hdTimeSpan = hdA - hdB;
			Assert.IsTrue(hdTimeSpan == timeSpan);
		}

		[Test]
		public void TestAdditionOverflow()
		{
			Assert.Throws<OverflowException>(delegate { var t = TimeSpan.MaxValue + TimeSpan.MaxValue; });
			Assert.Throws<OverflowException>(delegate { var t = TimeSpan.MinValue + TimeSpan.MinValue; });
			Assert.Throws<OverflowException>(delegate { var t = HdTimeSpan.MaxValue + HdTimeSpan.MaxValue; });
			Assert.Throws<OverflowException>(delegate { var t = HdTimeSpan.MinValue + HdTimeSpan.MinValue; });
			Assert.Throws<OverflowException>(delegate { var t = HdTimeSpan.MaxValue + new HdTimeSpan(1); });
			Assert.Throws<OverflowException>(delegate { var t = HdTimeSpan.MinValue + new HdTimeSpan(-1); });
		}

		[Test]
		public void TestSubtractionOverflow()
		{
			Assert.Throws<OverflowException>(delegate { var t = TimeSpan.MaxValue - TimeSpan.MinValue; });
			Assert.Throws<OverflowException>(delegate { var t = TimeSpan.MinValue - TimeSpan.MaxValue; });
			Assert.Throws<OverflowException>(delegate { var t = HdTimeSpan.MaxValue - HdTimeSpan.MinValue; });
			Assert.Throws<OverflowException>(delegate { var t = HdTimeSpan.MinValue - HdTimeSpan.MaxValue; });
			Assert.Throws<OverflowException>(delegate { var t = HdTimeSpan.MinValue - new HdTimeSpan(1); });
			Assert.Throws<OverflowException>(delegate { var t = HdTimeSpan.MaxValue - new HdTimeSpan(-1); });
		}


#pragma warning disable 1718
		[Test]
		public void TestEquals()
		{
			HdTimeSpan a1 = new HdTimeSpan(123);
			HdTimeSpan a2 = new HdTimeSpan(123);
			HdTimeSpan b = new HdTimeSpan(42);

			Assert.True(a1 == a1);
			Assert.False(a1 != a1);

			Assert.True(a1 == a2);
			Assert.False(a1 != a2);
			Assert.True(a1.Equals(a2));
			Assert.True(a2.Equals(a1));

			Assert.True(a1.Equals((Object)a1));
			Assert.True(a1.Equals((Object)a2));
			Assert.True(a2.Equals((Object)a1));
			Assert.True(((Object)a1).Equals((Object)a1));
			Assert.True(((Object)a1).Equals((Object)a2));
			Assert.True(((Object)a2).Equals((Object)a1));
			Assert.True(((Object)a1).Equals((Object)a2));
			Assert.True(((Object)a2).Equals((Object)a1));

			Assert.False(a1 == b);
			Assert.False(a2 == b);
			Assert.True(a1 != b);

			Assert.False(a1.Equals(b));
			Assert.False(a2.Equals(b));

			Assert.False(a1.Equals((Object)b));
			Assert.False(((Object)a1).Equals(b));
			Assert.False(((Object)a1).Equals((Object)b));
		}
#pragma warning restore 1718

		[Test]
		public void TestPrint()
		{
			TimeSpan ts = new TimeSpan(123, 45, 67, 9, 1234);
			HdTimeSpan hts = new HdTimeSpan(ts);
			Console.Out.WriteLine(ts);
			Console.Out.WriteLine(hts);
			Assert.AreEqual(ts.ToString(), hts.ToString("d.HH:mm:ss.fffffff"));
		}

		public void CheckPrint(String str, String fmt, HdTimeSpan ts, bool compareWithTimeSpan = false)
		{
			Assert.AreEqual(str, ts.ToString(fmt));
			if (compareWithTimeSpan)
				Assert.AreEqual(ts.TimeSpan.ToString(fmt), ((HdTimeSpan)(ts.TimeSpan)).ToString(fmt));
			//StringBuilder sb = new StringBuilder();
			//Assert.AreEqual(str, num.appendTo(sb, fmt).toString());
		}

		private void CheckParse(String from, String fmt, HdTimeSpan expected)
		{
			HdTimeSpan parsed = HdTimeSpan.Parse(from, fmt);
			if (!expected.Equals(parsed))
			{
				// Comparison is here to avoid problems with Formatter affecting tests for Parser
				Assert.AreEqual(expected, parsed);
				Assert.AreEqual(expected.ToString(), parsed.ToString());
			}

			Assert.AreEqual(expected.TotalNanoseconds, parsed.TotalNanoseconds);

		}

		private void CheckFormatFail(String from, String fmt, String msg)
		{
			try
			{
				HdTimeSpan.Parse(from, fmt);
			}
			catch (FormatError e)
			{
				if (e.Message.Contains(msg))
					return;
			}

			Assert.Fail("Was expected to throw");
		}

		// TODO: Convert to Test Cases (low priority)
		[Test]
		public void TestFormat()
		{
			// Plain numbers
			CheckPrint("34627623,.45634", "34627623,.45634", HdTimeSpan.FromSeconds(12), false);

			// Check quoted text
			CheckPrint("Abcmsy", "'Abcmsy'", HdTimeSpan.Zero);
			CheckPrint("00Abcmsy000", "00'Abcmsy'000", HdTimeSpan.Zero, false);
			CheckPrint("'Abc'msy", "'''Abc''msy'", HdTimeSpan.Zero, false);
			CheckPrint("0'0Abc''msy00'0", "0''0'Abc''''msy'00''0", HdTimeSpan.Zero, false);

			// Seconds
			CheckPrint("12", "s", HdTimeSpan.FromSeconds(12));
			CheckPrint("0", "s", HdTimeSpan.FromSeconds(0));
			CheckPrint("00", "ss", HdTimeSpan.FromSeconds(0));
			CheckPrint("005", "0ss", HdTimeSpan.FromSeconds(65));
			CheckPrint("000005", "ssssss", HdTimeSpan.FromSeconds(65));

			// Seconds & Fractions of Second. 'S' and 'f' are now synonyms
			CheckPrint("12.3", "s.S", HdTimeSpan.FromMilliseconds(12_300));
			CheckPrint("0.345", "s.SSS", HdTimeSpan.FromMicroseconds(345_000));
			CheckPrint("00.023", "ss.SSS", HdTimeSpan.FromMilliseconds(600_023));
			CheckPrint("05.123", "ss.SSS", HdTimeSpan.FromMilliseconds(65_123));
			CheckPrint("05.123000", "ss.SSSSSS", HdTimeSpan.FromMilliseconds(65_123));

			CheckPrint("05.0001", "ss.ffff", HdTimeSpan.FromMicroseconds(65_000_123));
			CheckPrint("05.00012", "ss.fffff", HdTimeSpan.FromMicroseconds(65_000_123));
			CheckPrint("05.000123", "ss.ffffff", HdTimeSpan.FromMicroseconds(65_000_123));
			CheckPrint("05.123000", "ss.ffffff", HdTimeSpan.FromNanoseconds(65_123_000_123L));
			CheckPrint("05.123000", "ss.ffffff", HdTimeSpan.FromNanoseconds(65_123_000_999L));
			CheckPrint("05.123000", "ss.ffffff", HdTimeSpan.FromNanoseconds(65_123_000_999L));
			CheckPrint("05.1230009", "ss.fffffff", HdTimeSpan.FromNanoseconds(65_123_000_999L));
			CheckPrint("05.12300012", "ss.ffffffff", HdTimeSpan.FromNanoseconds(65_123_000_123L));
			CheckPrint("05.123000123", "ss.fffffffff", HdTimeSpan.FromNanoseconds(65_123_000_123L));
			CheckPrint("05.000000123", "ss.fffffffff", HdTimeSpan.FromNanoseconds(65_000_000_123L));
			CheckPrint("5.000123000", "s.fffffffff", HdTimeSpan.FromNanoseconds(65_000_123_000L));

			// Minutes
			CheckPrint("5", "m", HdTimeSpan.FromMinutes(425));
			CheckPrint("7", "m", HdTimeSpan.FromSeconds(425));
			CheckPrint("05", "mm", HdTimeSpan.FromMinutes(425));
			CheckPrint("00005", "0mmmm", HdTimeSpan.FromMinutes(425));

			// Hours
			CheckPrint("5", "H", HdTimeSpan.FromHours(48 + 5));
			CheckPrint("4", "H", HdTimeSpan.FromMinutes(245));
			CheckPrint("07", "HH", HdTimeSpan.FromMinutes(425));
			CheckPrint("0007005", "0HHHmmm", HdTimeSpan.FromMinutes(425));
			CheckPrint("07:5.789", "HH:m.SSS", HdTimeSpan.FromMinutes(425).Add(HdTimeSpan.FromMilliseconds(789)));

			// Sign insertion
			CheckPrint("-12", "s", HdTimeSpan.FromSeconds(-12));
			CheckPrint("-0", "s", HdTimeSpan.FromMilliseconds(-1));
			CheckPrint("-05", "ss", HdTimeSpan.FromSeconds(-425));

			// "Advanced" sign insertion
			CheckPrint("-000005", "ssssss", HdTimeSpan.FromSeconds(-425));
			CheckPrint("-00000005", "00ssssss", HdTimeSpan.FromSeconds(-425));
			CheckPrint("Abc-00000005", "Abc00ssssss", HdTimeSpan.FromSeconds(-425));
			CheckPrint("Abcmsy-00000005", "'Abcmsy'00ssssss", HdTimeSpan.FromSeconds(-425));
			CheckPrint(";.:-000005", ";.:ssssss", HdTimeSpan.FromSeconds(-425));
			CheckPrint(";.:-00000005", ";.:00ssssss", HdTimeSpan.FromSeconds(-425));
		}

		[Test]
		public void TestParse()
		{
			CheckParse("-0000091024123456", "ddddddddddHHmmss", new HdTimeSpan(-91024, -12, -34, -56));

			CheckParse("5", "s", HdTimeSpan.FromSeconds(5));
			CheckParse("0", "s", HdTimeSpan.FromSeconds(0));
			CheckParse("005", "s", HdTimeSpan.FromSeconds(5));

			CheckParse("12:34:56", "H:m:s", new HdTimeSpan(12, 34, 56));
			CheckParse("12:34:56", "HH:mm:ss", new HdTimeSpan(12, 34, 56));

			CheckParse("1024T12:34:56", "dTH:m:s", new HdTimeSpan(1024, 12, 34, 56));
			CheckParse("1024T12:34:56", "d'T'H:m:s", new HdTimeSpan(1024, 12, 34, 56));

			// Unseparated fixed width fields
			CheckParse("123", "ss", HdTimeSpan.FromSeconds(12));
			CheckParse("123456", "HHmmss", new HdTimeSpan(12, 34, 56));
			CheckParse("1024123456", "ddddHHmmss", new HdTimeSpan(1024, 12, 34, 56));
			CheckParse("91024123456", "dddddHHmmss", new HdTimeSpan(91024, 12, 34, 56));
			CheckParse("0000091024123456", "ddddddddddHHmmss", new HdTimeSpan(91024, 12, 34, 56));
			CheckParse("101024123456", "ddddddHHmmss", new HdTimeSpan(101024, 12, 34, 56));

			// Sign
			CheckParse("-5", "s", HdTimeSpan.FromSeconds(-5));
			CheckParse("-1024T12:34:56", "dTH:m:s", new HdTimeSpan(-1024, -12, -34, -56));
			CheckParse("-123456", "HHmmss", new HdTimeSpan(-12, -34, -56));
			CheckParse("-0000091024123456", "ddddddddddHHmmss", new HdTimeSpan(-91024, -12, -34, -56));

			// Fractions (Only fixed length patterns supported)
			CheckParse("4.2", "s.f", HdTimeSpan.FromMilliseconds(4200));
			CheckParse("4.200000", "s.ffffff", HdTimeSpan.FromMilliseconds(4200));
			CheckParse("4.020", "s.ff", HdTimeSpan.FromMilliseconds(4020));
			CheckParse("4.200000000", "s.fffffffff", HdTimeSpan.FromMilliseconds(4200));
			CheckParse("4.000000002", "s.fffffffff", HdTimeSpan.FromNanoseconds(4000000002L));

			CheckParse("4.2", "s.S", HdTimeSpan.FromMilliseconds(4200));
			CheckParse("4.123", "s.SSS", HdTimeSpan.FromMilliseconds(4123));
		}

		[Test]
		public void testParseInvalidFormat()
		{
			CheckFormatFail("5 5", "s s", "Dup");
			CheckFormatFail("2002 2002", "dddd dddd", "Dup");
			CheckFormatFail("1 1", "d d", "Dup");
		}
	}
}

File: java/main/src/test/java/com/epam/deltix/hdtime/HdTimeSpanTest.java
/*
 * Copyright 2021 EPAM Systems, Inc
 *
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership. Licensed under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.epam.deltix.hdtime;

import org.junit.Assert;
import org.junit.Test;

import java.io.IOException;
import java.util.function.BiFunction;
import java.util.function.BinaryOperator;

public class HdTimeSpanTest {
    @Test
    @ValueTypeSuppressWarnings({"refArgs"})
    public void testNull() {
        Assert.assertNull(HdTimeSpan.fromUnderlying(HdDateTimeUtils.NULL_VALUE));
        Assert.assertNull(HdTimeSpan.fromUnderlying(HdDateTime.NULL_VALUE));
        Assert.assertNull(HdTimeSpan.NULL);

        HdTimeSpan nul = HdTimeSpan.NULL;
        Assert.assertNull(nul);
        Assert.assertEquals(nul, HdDateTime.fromUnderlying(HdTimeSpan.NULL_VALUE));
    }

    void checkPrint(String str, String fmt, HdTimeSpan num) throws IOException {
        Assert.assertEquals(str, num.toString(fmt));
        StringBuilder sb = new StringBuilder();
        Assert.assertEquals(str, num.appendTo(sb, fmt).toString());
    }

    private void checkFormatFail(String from, String fmt, String msg) throws ParseException {
        try {
            HdTimeSpan.parse(from, fmt);
        }
        catch (FormatError e) {
            if (e.getMessage().contains(msg))
                return;
        }

        Assert.fail("Was expected to throw");
    }

    @ValueTypeSuppressWarnings({"refArgs"})
    private void checkParse(String from, String fmt, HdTimeSpan expected) throws ParseException {
        HdTimeSpan parsed = HdTimeSpan.parse(from, fmt);
        if (!expected.equals(parsed)) {
            // Comparison is here to avoid problems with Formatter affecting tests for Parser
            Assert.assertEquals(expected, parsed);
            Assert.assertEquals(expected.toString(), parsed.toString());
        }

        Assert.assertEquals(HdTimeSpan.toUnderlying(expected), HdTimeSpan.toUnderlying(parsed));

    }

    @Test
    public void testPrint() throws IOException {

        // Plain numbers
        checkPrint("34627623,.45634", "34627623,.45634", HdTimeSpan.fromSeconds(12));

        // Check quoted text
        checkPrint("Abcmsy", "'Abcmsy'", HdTimeSpan.ZERO);
        checkPrint("00Abcmsy000", "00'Abcmsy'000", HdTimeSpan.ZERO);
        checkPrint("'Abc'msy", "'''Abc''msy'", HdTimeSpan.ZERO);
        checkPrint("0'0Abc''msy00'0", "0''0'Abc''''msy'00''0", HdTimeSpan.ZERO);

        // Seconds
        checkPrint("12", "s", HdTimeSpan.fromSeconds(12));
        checkPrint("0", "s", HdTimeSpan.fromSeconds(0));
        checkPrint("00", "ss", HdTimeSpan.fromSeconds(0));
        checkPrint("005", "0ss", HdTimeSpan.fromSeconds(65));
        checkPrint("000005", "ssssss", HdTimeSpan.fromSeconds(65));

        // Seconds & Fractions of Second. 'S' and 'f' are now synonyms
        checkPrint("12.3", "s.S", HdTimeSpan.fromMilliseconds(12_300));
        checkPrint("0.345", "s.SSS", HdTimeSpan.fromMicroseconds(345_000));
        checkPrint("00.023", "ss.SSS", HdTimeSpan.fromMilliseconds(600_023));
        checkPrint("05.123", "ss.SSS", HdTimeSpan.fromMilliseconds(65_123));
        checkPrint("05.123000", "ss.SSSSSS", HdTimeSpan.fromMilliseconds(65_123));

        checkPrint("05.0001", "ss.ffff", HdTimeSpan.fromMicroseconds(65_000_123));
        checkPrint("05.00012", "ss.fffff", HdTimeSpan.fromMicroseconds(65_000_123));
        checkPrint("05.000123", "ss.ffffff", HdTimeSpan.fromMicroseconds(65_000_123));
        checkPrint("05.123000", "ss.ffffff", HdTimeSpan.fromNanoseconds(65_123_000_123L));
        checkPrint("05.123000", "ss.ffffff", HdTimeSpan.fromNanoseconds(65_123_000_999L));
        checkPrint("05.123000", "ss.ffffff", HdTimeSpan.fromNanoseconds(65_123_000_999L));
        checkPrint("05.1230009", "ss.fffffff", HdTimeSpan.fromNanoseconds(65_123_000_999L));
        checkPrint("05.12300012", "ss.ffffffff", HdTimeSpan.fromNanoseconds(65_123_000_123L));
        checkPrint("05.123000123", "ss.fffffffff", HdTimeSpan.fromNanoseconds(65_123_000_123L));
        checkPrint("05.000000123", "ss.fffffffff", HdTimeSpan.fromNanoseconds(65_000_000_123L));
        checkPrint("5.000123000", "s.fffffffff", HdTimeSpan.fromNanoseconds(65_000_123_000L));

        // Minutes
        checkPrint("5", "m", HdTimeSpan.fromMinutes(425));
        checkPrint("7", "m", HdTimeSpan.fromSeconds(425));
        checkPrint("05", "mm", HdTimeSpan.fromMinutes(425));
        checkPrint("00005", "0mmmm", HdTimeSpan.fromMinutes(425));

        // Hours
        checkPrint("5", "H", HdTimeSpan.fromHours(48 + 5));
        checkPrint("4", "H", HdTimeSpan.fromMinutes(245));
        checkPrint("07", "HH", HdTimeSpan.fromMinutes(425));
        checkPrint("0007005", "0HHHmmm", HdTimeSpan.fromMinutes(425));
        checkPrint("07:5.789", "HH:m.SSS", HdTimeSpan.fromMinutes(425).add(HdTimeSpan.fromMilliseconds(789)));

        // Sign insertion
        checkPrint("-12", "s", HdTimeSpan.fromSeconds(-12));
        checkPrint("-0", "s", HdTimeSpan.fromMilliseconds(-1));
        checkPrint("-05", "ss", HdTimeSpan.fromSeconds(-425));

        // "Advanced" sign insertion
        checkPrint("-000005", "ssssss", HdTimeSpan.fromSeconds(-425));
        checkPrint("-00000005", "00ssssss", HdTimeSpan.fromSeconds(-425));
        checkPrint("Abc-00000005", "Abc00ssssss", HdTimeSpan.fromSeconds(-425));
        checkPrint("Abcmsy-00000005", "'Abcmsy'00ssssss", HdTimeSpan.fromSeconds(-425));
        checkPrint(";.:-000005", ";.:ssssss", HdTimeSpan.fromSeconds(-425));
        checkPrint(";.:-00000005", ";.:00ssssss", HdTimeSpan.fromSeconds(-425));
    }


    @Test
    public void testParse() throws IOException, ParseException {

        checkParse("5", "s", HdTimeSpan.fromSeconds(5));
        checkParse("0", "s", HdTimeSpan.fromSeconds(0));
        checkParse("005", "s", HdTimeSpan.fromSeconds(5));

        checkParse("12:34:56", "H:m:s", HdTimeSpan.newInstance(12, 34, 56));
        checkParse("12:34:56", "HH:mm:ss", HdTimeSpan.newInstance(12, 34, 56));

        checkParse("1024T12:34:56", "dTH:m:s", HdTimeSpan.newInstance(1024, 12, 34, 56));
        checkParse("1024T12:34:56", "d'T'H:m:s", HdTimeSpan.newInstance(1024, 12, 34, 56));

        // Unseparated fixed width fields
        checkParse("123", "ss", HdTimeSpan.fromSeconds(12));
        checkParse("123456", "HHmmss", HdTimeSpan.newInstance(12, 34, 56));
        checkParse("1024123456", "ddddHHmmss", HdTimeSpan.newInstance(1024,12, 34, 56));
        checkParse("91024123456", "dddddHHmmss", HdTimeSpan.newInstance(91024,12, 34, 56));
        checkParse("0000091024123456", "ddddddddddHHmmss", HdTimeSpan.newInstance(91024,12, 34, 56));
        checkParse("101024123456", "ddddddHHmmss", HdTimeSpan.newInstance(101024,12, 34, 56));

        // Sign
        checkParse("-5", "s", HdTimeSpan.fromSeconds(-5));
        checkParse("-1024T12:34:56", "dTH:m:s", HdTimeSpan.newInstance(-1024, -12, -34, -56));
        checkParse("-123456", "HHmmss", HdTimeSpan.newInstance(-12, -34, -56));
        checkParse("-0000091024123456", "ddddddddddHHmmss", HdTimeSpan.newInstance(-91024, -12, -34, -56));

        // Fractions (Only fixed length patterns supported)
        checkParse("4.2", "s.f", HdTimeSpan.fromMilliseconds(4200));
        checkParse("4.200000", "s.ffffff", HdTimeSpan.fromMilliseconds(4200));
        checkParse("4.020", "s.ff", HdTimeSpan.fromMilliseconds(4020));
        checkParse("4.200000000", "s.fffffffff", HdTimeSpan.fromMilliseconds(4200));
        checkParse("4.000000002", "s.fffffffff", HdTimeSpan.fromNanoseconds(4000000002L));

        checkParse("4.2", "s.S", HdTimeSpan.fromMilliseconds(4200));
        checkParse("4.123", "s.SSS", HdTimeSpan.fromMilliseconds(4123));
    }

    @Test
    public void testParseInvalidFormat() throws IOException, ParseException {
        checkFormatFail("5 5", "s s", "Dup");
        checkFormatFail("2002 2002", "dddd dddd", "Dup");
        checkFormatFail("1 1", "d d", "Dup");
    }


    @Test
    public void allMethods() throws ParseException, IOException {
        HdTimeSpan min = HdTimeSpan.MIN_VALUE;
        HdTimeSpan max = HdTimeSpan.MAX_VALUE;
        HdTimeSpan zero = HdTimeSpan.ZERO;

        HdTimeSpan.fromUnderlying(123456789);
        HdTimeSpan.fromUnderlying(123456789123456789L);
        HdTimeSpan.newInstance(0, 0,0);
        HdTimeSpan.newInstance(0, 0, 0,0);
        HdTimeSpan.newInstance(0, 0, 0,0, 0);

        HdTimeSpan.fromDays(0L);
        HdTimeSpan.fromHours(0L);
        HdTimeSpan.fromMinutes(0L);
        HdTimeSpan.fromSeconds(0L);
        HdTimeSpan.fromMilliseconds(0L);
        HdTimeSpan.fromMicroseconds(0L);
        HdTimeSpan.fromNanoseconds(0L);

        HdTimeSpan.fromDays(0.0);
        HdTimeSpan.fromHours(0.0);
        HdTimeSpan.fromMinutes(0.0);
        HdTimeSpan.fromSeconds(0.0);
        HdTimeSpan.fromMilliseconds(0.0);
        HdTimeSpan.fromMicroseconds(0.0);
        
        zero.getDays();
        zero.getHours();
        zero.getMinutes();
        zero.getSeconds();
        zero.getNanoseconds();

        zero.totalWeeks();
        zero.totalDays();
        zero.totalHours();
        zero.totalMinutes();
        zero.totalSeconds();
        zero.totalMilliseconds();
        zero.totalMicroseconds();
        zero.totalNanoseconds();

        HdTimeSpan a = HdTimeSpan.fromSeconds(-1), b = HdTimeSpan.fromSeconds(1);

        a.compareTo(b);
        HdTimeSpan.compare(a, b);
        HdTimeSpan.isLess(a, b);
        HdTimeSpan.isLessOrEqual(a, b);
        HdTimeSpan.isGreater(a, b);
        HdTimeSpan.isGreaterOrEqual(a, b);
        a.isLess(b);
        a.isLessOrEqual(b);
        a.isGreater(b);
        a.isGreaterOrEqual(b);
        a.isZero();
        a.isPositive();
        a.isNegative();
        a.roundTo(Resolution.MILLISECOND);
        a.roundTo(HdTimeSpan.fromMinutes(5));

        a.add(b);
        a.addUnchecked(b);
        a.subtract(b);
        a.subtractUnchecked(b);
        a.negate();
        a.duration();
        a.addDays(1);
        a.addHours(1);
        a.addMinutes(1);
        a.addSeconds(1);
        a.addMilliseconds(1);
        a.addMicroseconds(1);
        a.addNanoseconds(1);

        a.hashCode();
        a.equals("Nothing");
        a.equals(b);
        HdTimeSpan.equals(a, b);

        a.toString();
        a.toString("H:mm:ss,fffffffff");
        StringBuilder sb = new StringBuilder();
        a.appendTo(sb);
        a.appendTo(sb, "H:m:s");

        HdTimeSpan.parse("42 11:11:11.111111111");
        HdTimeSpan.parse("42", "d");
    }

    private void check(HdTimeSpan a, long b,
                       BiFunction<HdTimeSpan, Long, HdTimeSpan> f1,
                       BiFunction<HdTimeSpan, Long, HdTimeSpan> f2)  {

        HdTimeSpan x1 = HdTimeSpan.NULL, x2 = x1;
        Throwable e1 = null, e2 = null;
        try {
            x1 = f1.apply(a, b);
        }
        catch (Throwable e) {
            e1 = e;
        }

        try {
            x2 = f2.apply(a, b);
        }
        catch (Throwable e) {
            e2 = e;
        }

        if (e1 != null) {
            if (e2 != null) {
                if (e1.getClass() == e2.getClass())
                    return;

                Assert.fail(String.format("%s throws %s, while %s throws %s", f1, e1, f2, x2));
            }

            Assert.fail(String.format("%s throws %s, while %s returns %s", f1, e1, f2, x2));
        } else if (e2 != null) {
            Assert.fail(String.format("%s returns %s, while %s throws %s", f1, x1, f2, e2));
        }

        Assert.assertEquals(x1, x2);
    }


    @Test
    public void testAddXXBasic() {
        HdTimeSpan a = HdDateTime.now().subtract(HdDateTime.fromEpochMilliseconds(0));
        Assert.assertEquals(a.add(HdTimeSpan.fromDays(-123)), a.addDays(-123));
        Assert.assertEquals(a.add(HdTimeSpan.fromHours(321)), a.addHours(321));
        Assert.assertEquals(a.add(HdTimeSpan.fromMinutes(432)), a.addMinutes(432));
        Assert.assertEquals(a.add(HdTimeSpan.fromSeconds(-543)), a.addSeconds(-543));
        Assert.assertEquals(a.add(HdTimeSpan.fromMilliseconds(987654321)), a.addMilliseconds(987654321));
        Assert.assertEquals(a.add(HdTimeSpan.fromMicroseconds(-987654321)), a.addMicroseconds(-987654321));
        Assert.assertEquals(a.add(HdTimeSpan.fromNanoseconds(-123456789987654321L)), a.addNanoseconds(-123456789987654321L));
    }
}